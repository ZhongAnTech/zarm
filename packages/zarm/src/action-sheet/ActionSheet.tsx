import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Popup from '../popup';

export interface ActionSheetProps extends PropsType {
  prefixCls?: string;
  className?: string;
  safeIphoneX?: boolean;
}

export default class ActionSheet extends PureComponent<ActionSheetProps, {}> {
  static defaultProps: ActionSheetProps = {
    prefixCls: 'za-action-sheet',
    visible: false,
    spacing: false,
    actions: [],
    destroy: true,
  };

  renderActions = (action, index) => {
    const { prefixCls } = this.props;
    const actionCls = classnames(`${prefixCls}__item`, action.className, {
      [`${prefixCls}__item--${action.theme}`]: !!action.theme,
    });
    return (
      <div key={+index} className={actionCls} onClick={action.onClick}>
        {action.text}
      </div>
    );
  };

  renderCancel = () => {
    const { prefixCls, onCancel, cancelText, locale } = this.props;
    return (
      typeof onCancel === 'function' && (
        <div className={`${prefixCls}__cancel`}>
          <div className={`${prefixCls}__item`} onClick={onCancel}>
            {cancelText || locale!.cancelText}
          </div>
        </div>
      )
    );
  };

  render() {
    const { prefixCls, className, spacing, visible, onMaskClick, actions, destroy } = this.props;
    const cls = classnames(prefixCls, {
      [`${prefixCls}--spacing`]: spacing,
    });

    return (
      <Popup className={className} visible={visible} onMaskClick={onMaskClick} destroy={destroy}>
        <div className={cls}>
          <div className={`${prefixCls}__actions`}>{actions!.map(this.renderActions)}</div>
          {this.renderCancel()}
        </div>
      </Popup>
    );
  }
}
