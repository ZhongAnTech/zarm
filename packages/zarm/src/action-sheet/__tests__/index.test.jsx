import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ActionSheet from '../index';
import zhCN from '../locale/zh_CN';
import enUS from '../locale/en_US';

describe('ActionSheet', () => {
  const props = {
    visible: true,
    actions: [
      {
        text: '操作一',
        onClick: jest.fn(),
      },
      {
        text: '操作二',
        onClick: jest.fn(),
      },
      {
        theme: 'danger',
        text: '操作三',
        onClick: jest.fn(),
      },
    ],
    onMaskClick: jest.fn(),
    onCancel: jest.fn(),
  };
  it('renders correctly', () => {
    const wrapper = mount(<ActionSheet {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('locale', () => {
    expect(zhCN.cancelText).toEqual('取消');
    expect(enUS.cancelText).toEqual('Cancel');
  });
});
