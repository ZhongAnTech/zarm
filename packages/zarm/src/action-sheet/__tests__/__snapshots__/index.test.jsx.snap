// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet renders correctly 1`] = `
<ForwardRef(forwardRef)
  actions={
    Array [
      Object {
        "onClick": [MockFunction],
        "text": "操作一",
      },
      Object {
        "onClick": [MockFunction],
        "text": "操作二",
      },
      Object {
        "onClick": [MockFunction],
        "text": "操作三",
        "theme": "danger",
      },
    ]
  }
  onCancel={[MockFunction]}
  onMaskClick={[MockFunction]}
  visible={true}
>
  <ConfigReceiver
    actions={
      Array [
        Object {
          "onClick": [MockFunction],
          "text": "操作一",
        },
        Object {
          "onClick": [MockFunction],
          "text": "操作二",
        },
        Object {
          "onClick": [MockFunction],
          "text": "操作三",
          "theme": "danger",
        },
      ]
    }
    forwardedRef={null}
    onCancel={[MockFunction]}
    onMaskClick={[MockFunction]}
    visible={true}
  >
    <ActionSheet
      actions={
        Array [
          Object {
            "onClick": [MockFunction],
            "text": "操作一",
          },
          Object {
            "onClick": [MockFunction],
            "text": "操作二",
          },
          Object {
            "onClick": [MockFunction],
            "text": "操作三",
            "theme": "danger",
          },
        ]
      }
      destroy={true}
      locale={
        Object {
          "cancelText": "取消",
        }
      }
      localeCode="zh-CN"
      onCancel={[MockFunction]}
      onMaskClick={[MockFunction]}
      prefixCls="za-action-sheet"
      spacing={false}
      visible={true}
    >
      <ForwardRef(ForwardedRefComp)
        destroy={true}
        onMaskClick={[MockFunction]}
        visible={true}
      >
        <WarnIfDeprecatedComp
          destroy={true}
          forwardedRef={null}
          onMaskClick={[MockFunction]}
          visible={true}
        >
          <Popup
            destroy={true}
            onMaskClick={[MockFunction]}
            prefixCls="za-popup"
            visible={true}
          >
            <Portal
              animationDuration={200}
              animationType="fade"
              destroy={true}
              direction="bottom"
              handlePortalUnmount={[Function]}
              mask={true}
              maskType="normal"
              onMaskClick={[MockFunction]}
              prefixCls="za-popup"
              visible={true}
            >
              <Trigger
                disabled={false}
                onClose={[Function]}
                visible={true}
              >
                <Portal
                  containerInfo={
                    <div
                      class="za-popup-container"
                    >
                      <div
                        class="za-mask za-fade-enter za-mask--normal"
                        style="animation-duration: 200ms;"
                      />
                      <div
                        class="za-popup__wrapper"
                        data-width="0"
                        role="dialog"
                      >
                        <div
                          class="za-popup za-popup--bottom za-popup--show"
                          data-width="0"
                          role="document"
                          style="transition-duration: 200ms; transition-property: transform;"
                        >
                          <div
                            class="za-action-sheet"
                          >
                            <div
                              class="za-action-sheet__actions"
                            >
                              <div
                                class="za-action-sheet__item"
                              >
                                操作一
                              </div>
                              <div
                                class="za-action-sheet__item"
                              >
                                操作二
                              </div>
                              <div
                                class="za-action-sheet__item za-action-sheet__item--danger"
                              >
                                操作三
                              </div>
                            </div>
                            <div
                              class="za-action-sheet__cancel"
                            >
                              <div
                                class="za-action-sheet__item"
                              >
                                取消
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  }
                >
                  <Mask
                    className="za-fade-enter"
                    prefixCls="za-mask"
                    style={
                      Object {
                        "WebkitAnimationDuration": "200ms",
                        "animationDuration": "200ms",
                      }
                    }
                    type="normal"
                    visible={true}
                  >
                    <div
                      className="za-mask za-fade-enter za-mask--normal"
                      style={
                        Object {
                          "WebkitAnimationDuration": "200ms",
                          "animationDuration": "200ms",
                        }
                      }
                    />
                  </Mask>
                  <div
                    className="za-popup__wrapper"
                    data-width={0}
                    onClick={[Function]}
                    role="dialog"
                    style={Object {}}
                  >
                    <div
                      className="za-popup za-popup--bottom"
                      data-width={0}
                      role="document"
                      style={
                        Object {
                          "WebkitTransitionDuration": "200ms",
                          "WebkitTransitionProperty": "transform",
                          "transitionDuration": "200ms",
                          "transitionProperty": "transform",
                        }
                      }
                    >
                      <div
                        className="za-action-sheet"
                      >
                        <div
                          className="za-action-sheet__actions"
                        >
                          <div
                            className="za-action-sheet__item"
                            key="0"
                            onClick={[MockFunction]}
                          >
                            操作一
                          </div>
                          <div
                            className="za-action-sheet__item"
                            key="1"
                            onClick={[MockFunction]}
                          >
                            操作二
                          </div>
                          <div
                            className="za-action-sheet__item za-action-sheet__item--danger"
                            key="2"
                            onClick={[MockFunction]}
                          >
                            操作三
                          </div>
                        </div>
                        <div
                          className="za-action-sheet__cancel"
                        >
                          <div
                            className="za-action-sheet__item"
                            onClick={[MockFunction]}
                          >
                            取消
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </Portal>
              </Trigger>
            </Portal>
          </Popup>
        </WarnIfDeprecatedComp>
      </ForwardRef(ForwardedRefComp)>
    </ActionSheet>
  </ConfigReceiver>
</ForwardRef(forwardRef)>
`;
